File: web/src/verify/diff.ts
// This mirrors internal/diff/diff.go field-for-field. It only accepts report
// shapes whose attested content has already been independently verified by
// verifier.ts; the API's history response is transport, never the source of
// truth for a displayed change.
import type { AttestedContent, CheckOutput, CheckResult } from "./types";

export type DiffEventKind = "status_changed" | "new_finding" | "resolved_finding";
export type DiffSeverity = "regression" | "neutral" | "improvement";

export interface DiffEvent {
  kind: DiffEventKind;
  checkId: string;
  checkTitle: string;
  accountId: string;
  previousStatus?: string;
  currentStatus?: string;
  finding?: string;
  severity: DiffSeverity;
}

function accountResults(snapshot: AttestedContent, check: CheckOutput): Record<string, CheckResult> {
  return Object.keys(check.accounts ?? {}).length > 0 ? check.accounts! : { [snapshot.account_id]: check.result };
}

function setDifference(left: string[] | null | undefined, right: string[] | null | undefined): string[] {
  const prior = new Set(right ?? []);
  return [...new Set(left ?? [])].filter((value) => !prior.has(value)).sort();
}

function severity(kind: DiffEventKind, previousStatus?: string, currentStatus?: string): DiffSeverity {
  if (kind === "new_finding") return "regression";
  if (kind === "resolved_finding") return "improvement";
  if (currentStatus === "fail" || currentStatus === "error") return "regression";
  if (previousStatus === "fail" || previousStatus === "error") return "improvement";
  return "neutral";
}

/** Deterministically compare only controls and accounts both reports observed. */
export function diffReports(previous: AttestedContent, current: AttestedContent): DiffEvent[] {
  const events: DiffEvent[] = [];
  for (const [checkId, currentCheck] of Object.entries(current.checks)) {
    const previousCheck = previous.checks[checkId];
    if (!previousCheck) continue;
    const oldAccounts = accountResults(previous, previousCheck);
    const newAccounts = accountResults(current, currentCheck);
    for (const [accountId, currentResult] of Object.entries(newAccounts)) {
      const previousResult = oldAccounts[accountId];
      if (!previousResult) continue;
      if (previousResult.status !== currentResult.status) {
        events.push({
          kind: "status_changed", checkId, checkTitle: currentCheck.title, accountId,
          previousStatus: previousResult.status, currentStatus: currentResult.status,
          severity: severity("status_changed", previousResult.status, currentResult.status),
        });
      }
      for (const finding of setDifference(currentResult.findings, previousResult.findings)) {
        events.push({ kind: "new_finding", checkId, checkTitle: currentCheck.title, accountId, finding, severity: "regression" });
      }
      for (const finding of setDifference(previousResult.findings, currentResult.findings)) {
        events.push({ kind: "resolved_finding", checkId, checkTitle: currentCheck.title, accountId, finding, severity: "improvement" });
      }
    }
  }
  const rank: Record<DiffSeverity, number> = { regression: 0, neutral: 1, improvement: 2 };
  return events.sort((a, b) =>
    rank[a.severity] - rank[b.severity] || a.checkId.localeCompare(b.checkId) ||
    a.accountId.localeCompare(b.accountId) || a.kind.localeCompare(b.kind) || (a.finding ?? "").localeCompare(b.finding ?? ""),
  );
}

File: internal/diff/diff.go
// Package diff compares two attested report snapshots. It deliberately has
// no database or HTTP dependency: the exact same transition rules are mirrored
// in web/src/verify/diff.ts and run in the buyer's browser after both inputs
// have independently passed Nitro chain, signature, hash, and PCR checks.
package diff

import (
	"sort"

	"github.com/Georgy03/zerodock/internal/checks"
	"github.com/Georgy03/zerodock/internal/report"
)

type EventKind string

const (
	StatusChanged   EventKind = "status_changed"
	NewFinding      EventKind = "new_finding"
	ResolvedFinding EventKind = "resolved_finding"
)

// Event is one reproducible change between two snapshots. FirstObservedAt is
// intentionally not here: it is the current verdict's separately attested
// document timestamp, added only at presentation time by the caller.
type Event struct {
	Kind           EventKind
	CheckID        string
	CheckTitle     string
	AccountID      string
	PreviousStatus string
	CurrentStatus  string
	Finding        string
}

// Reports compares per-check, per-account results that exist in both
// snapshots. It reports status transitions plus findings added or resolved.
// A newly introduced check is not automatically a regression: older signed
// reports could not have evaluated code that did not exist yet. Likewise an
// account that appears or disappears is handled by scope.Detect, not disguised
// as a control failure.
func Reports(previous, current report.AttestedContent) []Event {
	var events []Event
	for checkID, currentCheck := range current.Checks {
		previousCheck, exists := previous.Checks[checkID]
		if !exists {
			continue
		}
		previousAccounts := accountResults(previous, previousCheck)
		currentAccounts := accountResults(current, currentCheck)
		for accountID, currentResult := range currentAccounts {
			previousResult, exists := previousAccounts[accountID]
			if !exists {
				continue
			}
			if previousResult.Status != currentResult.Status {
				events = append(events, Event{
					Kind:           StatusChanged,
					CheckID:        checkID,
					CheckTitle:     currentCheck.Title,
					AccountID:      accountID,
					PreviousStatus: previousResult.Status,
					CurrentStatus:  currentResult.Status,
				})
			}
			for _, finding := range setDifference(currentResult.Findings, previousResult.Findings) {
				events = append(events, Event{Kind: NewFinding, CheckID: checkID, CheckTitle: currentCheck.Title, AccountID: accountID, Finding: finding})
			}
			for _, finding := range setDifference(previousResult.Findings, currentResult.Findings) {
				events = append(events, Event{Kind: ResolvedFinding, CheckID: checkID, CheckTitle: currentCheck.Title, AccountID: accountID, Finding: finding})
			}
		}
	}
	sort.Slice(events, func(i, j int) bool {
		left, right := severity(events[i]), severity(events[j])
		if left != right {
			return left < right
		}
		if events[i].CheckID != events[j].CheckID {
			return events[i].CheckID < events[j].CheckID
		}
		if events[i].AccountID != events[j].AccountID {
			return events[i].AccountID < events[j].AccountID
		}
		if events[i].Kind != events[j].Kind {
			return events[i].Kind < events[j].Kind
		}
		return events[i].Finding < events[j].Finding
	})
	return events
}

func accountResults(snapshot report.AttestedContent, check report.CheckOutput) map[string]checks.Result {
	if len(check.Accounts) > 0 {
		return check.Accounts
	}
	return map[string]checks.Result{snapshot.AccountID: check.Result}
}

func setDifference(left, right []string) []string {
	rightSet := make(map[string]struct{}, len(right))
	for _, value := range right {
		rightSet[value] = struct{}{}
	}
	seen := make(map[string]struct{}, len(left))
	var out []string
	for _, value := range left {
		if _, duplicate := seen[value]; duplicate {
			continue
		}
		seen[value] = struct{}{}
		if _, exists := rightSet[value]; !exists {
			out = append(out, value)
		}
	}
	sort.Strings(out)
	return out
}

// severity orders regressions before neutral changes and improvements, so a
// buyer sees the things needing attention first without relying on map order.
func severity(event Event) int {
	if event.Kind == NewFinding {
		return 0
	}
	if event.Kind == ResolvedFinding {
		return 2
	}
	if event.CurrentStatus == checks.StatusFail || event.CurrentStatus == checks.StatusError {
		return 0
	}
	if event.PreviousStatus == checks.StatusFail || event.PreviousStatus == checks.StatusError {
		return 2
	}
	return 1
}
